freac, BoCA, smooth: new port

Janosch opened the pull request:
#### Description
New port for [fre:ac](https://www.freac.org/) audio converter and its dependencies [BoCA](https://github.com/enzo1982/BoCA) and [smooth](https://github.com/enzo1982/smooth).

###### Tested on
<!-- Triple-click and copy the next line and paste it into your shell. It will copy your OS and Xcode version to the clipboard. Paste it here replacing this section.
printf "%s\n" "macOS $(sw_vers -productVersion) $(sw_vers -buildVersion)" "$(xcodebuild -version|awk 'NR==1{x=$0}END{print x" "$NF}')"|tee /dev/tty|pbcopy
-->
macOS 11.1 20C69
Xcode 12.3.0.0.1.1607026830

###### Verification <!-- (delete not applicable items) -->
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [ ] referenced existing tickets on Trac with full URL? 
- [x] checked your Portfile with `port lint`?
- [ ] tried existing tests with `sudo port test`?
- [x] tried a full install with `sudo port -vst install`?
- [x] tested basic functionality of all binary files?

<!-- Use "skip notification" (surrounded with []) to avoid notifying maintainers -->

Janosch commented:
The azure pipeline fails because it is trying to install the ports in the wrong order. The dependency graph for freac is 

freac -> BoCA -> smooth 

but the script tries to install first BoCA then smooth.

Can this be ignored or do I really have to submit the individual ports one after another?

reneeotten commented on audio/freac/Portfile:
```diff
@@ -0,0 +1,45 @@
+categories          audio
+platforms           darwin
+license             GPL-2
+maintainers         {@janosch mailbox.org:janosch1}
```
can `openmaintainer` be added here?

Janosch replied:
Done.

reneeotten reviewed (changes requested):
@Janosch - thanks for the new ports! I just had a cursory look and listed some things that I noticed. I see that there is quite a bit going on in the `post-destroot` phase of these ports... I haven't tried to build these ports and am not sure whether all of that is required and/or cannot be done during the build phase. So a review from someone else is likely needed here before merging.

Janosch commented on aqua/smooth/Portfile:
```diff
@@ -0,0 +1,60 @@
+categories          aqua
+platforms           darwin
+license             ArtisticLicense-2
+maintainers         {@janosch mailbox.org:janosch1}
```
Yes. Done.

Janosch commented on aqua/smooth/files/patch-fix-libxml2-includes.diff:
```diff
@@ -0,0 +1,11 @@
+ 	MYCCOPTS += -I"$(SRCDIR)"/$(SMOOTH_PATH)/include/support/libxml2
+ else ifeq ($(BUILD_OSX),True)
+-	MYCCOPTS += -I$(shell xcodebuild -sdk macosx -version | grep "^Path: " | head -n 1 | tail -c +7)/usr/include/libxml2
++	MYCCOPTS += -I/opt/local/include/libxml2
```
Got it. I am using the prefix variable instead, which is already present in the Makefiles.

Janosch commented on audio/BoCA/Portfile:
```diff
@@ -0,0 +1,54 @@
+categories          audio
+platforms           darwin
+license             GPL-2
+maintainers         {@janosch mailbox.org:janosch1}
```
Done.

Janosch commented on aqua/smooth/Portfile:
```diff
@@ -0,0 +1,60 @@
+
+categories          aqua
+platforms           darwin
+license             ArtisticLicense-2
```
Done.

Janosch commented:
> @Janosch - thanks for the new ports! I just had a cursory look and listed some things that I noticed. I see that there is quite a bit going on in the `post-destroot` phase of these ports... I haven't tried to build these ports and am not sure whether all of that is required and/or cannot be done during the build phase. So a review from someone else is likely needed here before merging.

I fixed the portfiles according to your comments. Changing the references and IDs of the dylibs could also be done by patching the Makefile, but doing it afterwards with install_tool_name is more robust in regard of future changes to the Makefile. 

Furthermore, I removed the git dependency and also removed _--environment-overrides_ from build and destroot args. Both are leftovers which are not needed anymore.

mf2k commented on aqua/smooth/Portfile:
```diff
@@ -0,0 +1,60 @@
+                    \n* a libxml2 based XML parser\
+homepage            http://www.smooth-project.org/
+
+depends_lib         port:git \
+                    port:bzip2 \
+                    port:curl \
+                    port:fribidi \
+                    port:gtk3 \
+                    port:jpeg \
+                    port:libxml2
```
Please sort these alphabetically

Janosch replied:
Done. I also removed git as it is not a dependency. I added this in the first place because upstream is a github project, but now I realized that I dont use git to fetch the sources.

Janosch replied:
Done. I also removed git because it turned out that it is not a dependency.

mf2k commented on aqua/smooth/Portfile:
```diff
@@ -0,0 +1,60 @@
+                    port:curl \
+                    port:fribidi \
+                    port:gtk3 \
+                    port:jpeg \
```
We are no longer using jpeg and have switched to libjpeg-turbo. This is how it should be written now:

`path:include/turbojpeg.h:libjpeg-turbo`

Janosch replied:
Done.

mf2k commented on audio/boca/Portfile:
```diff
@@ -0,0 +1,53 @@
+                    port:uriparser \
+                    port:smooth
+
+use_configure       no
```
Because of this, you need to review and implement [UsingTheRightCompiler](https://trac.macports.org/wiki/UsingTheRightCompiler)

reneeotten replied:
that does apply to all three ports here, you might want to use the `makefile` PG as that will likely set some of the things for you correctly. But yes, Frank is right: please make sure these ports all use the correct compiler.

Janosch replied:
@mf2k Ok, I decided to remove `use_configure no` then. The configure phase does not create any makefiles but if it helps preserving the compiler selection I will keep the configure phase.

@reneeotten I added the makefile PG to all ports. The only feature I use is `makefile.prefix_name` though.

reneeotten replied:
if you use the `makefile` PG then it will automatically set `use_configure no`; you cannot run `./configure` since it doesn't exist for these ports. It's just that the `makefile` PG attempts to set the compilers and flags correctly so that if there is a "standard" Makefile that they can be used.

Janosch replied:
Ok after reading through makefile-1.0.tcl I think I got it. I am marking this point as resolved as the makefile PG will set the compiler envs the "macports way".

mf2k commented on aqua/smooth/Portfile:
```diff
@@ -0,0 +1,60 @@
+# After building smooth, references to libsmooth and the id of libsmooth
+# are not pointing to ${prefix}/lib. We are fixing this here.
+post-destroot       {
+                      system "install_name_tool -change libsmooth-${major}.${minor}.${compat}.dylib \
```
I'm surprised this works. You should use an absolute path to install_name_tool.

Janosch replied:
Hmm, I dont understand. The usage of install_name_tool is 
`install_name_tool [-change old new] input`
I am using absoluth paths to reference the input and to specify the new install name, but for `old` I just have to use the current install name. If I specify something else here install_name_tool will fail because it does not know what to change.

Janosch replied:
Done.

mf2k commented on aqua/smooth/Portfile:
```diff
@@ -0,0 +1,60 @@
+                      ${prefix}/lib/libsmooth-${major}.${minor}.${compat}.dylib \
+                      ${destroot}${prefix}/bin/smooth-translator"
+
+                      system "install_name_tool -id ${prefix}/lib/libsmooth-${major}.${minor}.${compat}.dylib \
```
Same here.

Janosch replied:
Done.

Janosch commented:
Thanks for the reviews @mf2k and @reneeotten. I resolved the open issues.

reneeotten commented:
> Thanks for the reviews @mf2k and @reneeotten. I resolved the open issues.

it does build for me locally, but I don't really know how to use any of it so I couldn't test if it actually works... 

Again, I am not sure whether all the stuff in the `post-destroot` is the best/recommended way of accomplishing this or if it could be done somehow in the `build` phase by changing the Makefile. But that's a bit out of my expertise (perhaps @kencu can give offer his insight?).

One comment remaining from me: since these are three new ports, there should be one commit for each port. If you can take care of that I'll merge this tomorrow unless others have comments/suggestions here. If there are things to be fixed I'm sure you'll be happy to take care of that later!

Janosch commented:
@reneeotten Well for testing you just need to start the freac app, for a more thorough test maybe insert an audio CD and rip it.

I will have a look if I can alter the install name of the dynamic libraries in the Makefile instead of in post-destroot phase, but I dont want to spend too much time on that.

I can make three seperate PRs for this but I'd like to point out that seperating those three ports in seperate PRs makes it really cumbersome for developing and reviewing. I will consider adding smooth and boca as subports of freac instead, as boca and smooth will almost certainly only be used by freac now and in the future.

mf2k commented:
These portfiles are different enough that leaving them separate seems to make the most sense in this case. sub-ports should be used to remove duplicate code in similar portfiles. Please do make the commits separate for each as @reneeotten requested.

Janosch commented:
@reneeotten @mf2k I did split up the commit into three commits one for each port. Also, I added patches which fixes the dylib install names for the _smooth_ and _boca_ port. Therefore, calling `install_name_tool` in post-destroot became obsolete for those ports.

I hope this is ready to merge now. If you still see issues I should fix let me know.

Janosch commented:
In the latest push I just did some reformatting: giving post-destroot in port freac the correct indent and moving checksums further up (in most ports they are not at the bottom but rather somwhere after the description). Also, I added some line breaks.

reneeotten commented:
thanks @Janosch - in it goes!
